compiler/rustc_ast_lowering/src/expr.rs

```diff
@@ -22,9 +24,28 @@ use super::errors::{
 use super::{
     GenericArgsMode, ImplTraitContext, LoweringContext, ParamMode, ResolverAstLoweringExt,
 };
-use crate::errors::YieldInClosure;
+use crate::errors::{InvalidLegacyConstGenericArg, YieldInClosure};
 use crate::{fluent_generated, FnDeclKind, ImplTraitPosition};
 
+struct WillCreateDefIdsVisitor {}
+
+impl<'v> rustc_ast::visit::Visitor<'v> for WillCreateDefIdsVisitor {
+    type Result = ControlFlow<()>;
+
+    fn visit_item(&mut self, _: &'v Item) -> Self::Result {
+        ControlFlow::Break(())
+    }
+
+    fn visit_expr(&mut self, ex: &'v Expr) -> Self::Result {
+        match ex.kind {
+            ExprKind::Gen(..) | ExprKind::ConstBlock(..) | ExprKind::Closure(..) => {
```

Comment:
maybe break with the value of `ControlFlow<Span>` so you can point at the span of the closure/gen block/etc specifically, rather than the whole expr?